### PR TITLE
Pagination

### PR DESCRIPTION
**Pull request by mcmillanator**
Pure ruby pagination.  This is not exactly what we did in class.  In class we used JS to show and hide elements to create the pages.  Here I am using SQL limit and offset to achieve pagination.  I will be mixing this with JS when I have the time.  I believe that both this method and JS should be used in combination.  Using one or the other has a trade-off between resources used and results.

Using pure ruby I am sending a request to the server for every single page.
This is bad because you don't want to send so many request to the server.
This is good because you don't want to load thousands of rows at once.

Using JS I am sending a single request to the server then hiding or showing data for every page. 
This is bad because you don't want to load thousands of rows at once.
This is good because you don't wan to send many requests to the server.

Using both methods you can load a limited number of rows from the database.  Then you can decide how many rows to send at one time.  You can also use JS to create the pages and then every X pages send a new request to the server for the next set of data.

### CODE REVIEW

File: app/controllers/items_controller.rb [hunk 1]
```diff
@@ -1,9 +1,42 @@
 class ItemsController < ApplicationController
+  
+  # TODO add pagination
```
Comment (mcmillanator) [inline app/controllers/items_controller.rb]: TODO remove this TODO

File: app/controllers/items_controller.rb [hunk 2]
```diff
@@ -1,9 +1,42 @@
...
+    i = 0
+    @pages = [0]
+    while i < total-1 do
+      i += 1
+      if i % 10 == 0
+        @pages.push(i)
+      end
+    end
+    @items = @current_customer.items.limit(10).offset(@pages[page])
+    # binding.pry
```
Comment (mcmillanator) [inline app/controllers/items_controller.rb]: TODO remove comment

File: app/views/items/pagination.html.erb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<div>
```
Comment (mcmillanator) [inline app/views/items/pagination.html.erb]: TODO remove this file.  It is not being used.

File: config/routes.rb [hunk 4]
```diff
@@ -3,11 +3,10 @@
...
   resources :sessions
-  get   '/login',   to: 'sessions#new'
-  post  '/login',   to: 'sessions#create'
-  get   '/logout',  to: 'sessions#destroy'
+  resources :items
 
-  # get '/signup', to: 'customers#create'
-
-  # put '/new', to: 'customers#create'
+  get   '/customer/items',           to: 'items#index'
```
Comment (mcmillanator) [inline config/routes.rb]: TODO fix this spacing

File: app/assets/javascripts/pagination.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            row += item.name
+            row += "</li>" 
+            row += "<li> <strong>ID</strong>: "
+            row += item.id
+            row += "</li>" 
+            row += "<li> <strong>Category</strong>: "
+            row += item.category
+            row += "</li>"
+            row += "<li> <strong>Price</strong>: $"
+            row += item.price_in_cents / 100
```
Comment (rpearce) [inline app/assets/javascripts/pagination.js]: if you're going to do it this way, don't forget your semi-colons

File: app/controllers/items_controller.rb [hunk 6]
```diff
@@ -1,9 +1,53 @@
...
+    @page = params[:page].to_i
+    @next = @page + 1
+    @previous = @page - 1
+    if @page == 0
+      current_page = 0
+    else
+      current_page = @page - 1
+    end
+    total = @current_customer.items.count
+    limit = 10
```
Comment (rpearce) [inline app/controllers/items_controller.rb]: I don't think you need to do all this logic... I believe you can just do `.limit(10)` on an `ActiveRecord` query

File: app/controllers/items_controller.rb [hunk 7]
```diff
@@ -1,9 +1,53 @@
...
+      i += 1
+      if i % limit == 0
+        @pages.push(i)
+      end
+    end
+    @count = @pages.count
+    @items = @current_customer.items.limit(limit).offset(@pages[current_page])
+    respond_to do |format|
+      format.html {render :items}
+      format.json {render json: {items: @items, total: @count, current_page: @page} }
```
Comment (rpearce) [inline app/controllers/items_controller.rb]: don't forget about the option to use jbuilder

File: app/views/items/index.html.erb [hunk 8]
```diff
@@ -1,15 +1,28 @@
...
+  <% end %>
+</div>
+<nav class="page">
+  <%= link_to "First", "/1" if @page > 1 %>
+  <%= link_to "Previous", "/?page=#{@previous}" if @previous > 0 %>
+  <% @pages.each_with_index do |p, i| %>
+    <% i += 1 %>
+    <%= link_to "Page #{i}", "/?page=#{i}" %>
+  <% end %>
+  <%= link_to "Next", "/?page=#{@next}" if @next < @count %>
```
Comment (rpearce) [inline app/views/items/index.html.erb]: good news! I believe you can actually do it like this
```haml
<%= link_to 'Next', root_path(page: @next) if @next < @count %>
```

File: config/routes.rb [hunk 9]
```diff
@@ -3,11 +3,11 @@
...
+  resources :items
 
-  # get '/signup', to: 'customers#create'
-
-  # put '/new', to: 'customers#create'
+  get   '/customer/items',  to: 'items#index'
+  get   '/:page',           to: 'items#index'
+  get   '/login',           to: 'sessions#new'
+  post  '/login',           to: 'sessions#create'
+  get   '/logout',          to: 'sessions#destroy'
```
Comment (rpearce) [inline config/routes.rb]: I recommend adding an `, as: :destroy_session` so that you can use `destroy_session_path`. This could work with all of them

**Pull request merged**